src/plugins/auto-moderation-plugin.ts

```diff
@@ -0,0 +1,161 @@
+import fs from 'node:fs'
+
+import type { PluginContext, PluginInterface } from '../common/plugins.js'
+
+import { EventType, InstanceType, PunishmentType, Severity, ChannelType } from 'src/common/application-event.js'
+import { antiSpamString } from 'src/util/shared-util.js'
+
+enum HeatType {
+  SPAM = 'spam',
+  PROFANITY = 'profanity'
+}
+
+type HeatScore = number
+
+type HeatsTable = Record<HeatType, HeatScore>
+type WarnsTable = Record<HeatType, boolean>
+type InfractionsTable = Record<HeatType, number>
+
+type HeatUser = string
+
+interface InfractionCase {
+  userName: string
+  userUuid: string | undefined
+  userDiscordId: number | undefined
+  infractions: InfractionsTable
+  lastInfraction: number
+}
+
+const heats: Record<HeatUser, HeatsTable> = {}
+const warns: Record<HeatUser, WarnsTable> = {}
+let infractions: InfractionCase[] = []
+
+setInterval(() => {
+  reduceHeats()
+}, 1000)
+
+function saveConfig(path: string): void {
+  const dataRaw = JSON.stringify(infractions, undefined, 4)
+  fs.writeFileSync(path, dataRaw, { encoding: 'utf8' })
+}
+
+function reduceInfractions() {
+  for (const infractionCase of infractions) {
+    if (infractionCase.lastInfraction <= Date.now() - 7 * 24 * 60 * 60 * 1000) {
+      infractions = infractions.filter((infraction) => infraction.userName !== infractionCase.userName)
+    }
+  }
+}
+
+function reduceHeats() {
+  for (const [heatUser, heatsTable] of Object.entries(heats)) {
+    // @ts-expect-error prefer-const
+    for (let [heatType, heatScore] of Object.entries(heatsTable)) {
+      switch (heatType) {
+        case HeatType.SPAM: {
+          heatScore -= 0.05
+          break
+        }
+        case HeatType.PROFANITY: {
+          heatScore -= 0.01
+          break
+        }
+      }
+      // @ts-expect-error no-dynamic-delete
+      if (heatScore <= 0) delete heatsTable[heatType as unknown as HeatType]
+      else heatsTable[heatType as unknown as HeatType] = heatScore
+    }
+
+    // @ts-expect-error no-dynamic-delete
+    if (Object.keys(heatsTable).length === 0) delete heats[heatUser], delete warns[heatUser]
+  }
+}
+
+export default {
+  onRun(context: PluginContext): void {
+    const infractionFilePath = context.application.getConfigFilePath('infractions.json')
+    infractions.join(JSON.parse(fs.readFileSync(infractionFilePath, 'utf8')))
```

Comment:
Separate each action into its own line and don't stack functions on each other like this.